libc/config/baremetal/arm/entrypoints.txt

```diff
@@ -19,6 +19,7 @@ set(TARGET_LIBC_ENTRYPOINTS
 
     # compiler entrypoints (no corresponding header)
     libc.src.compiler.__stack_chk_fail
+    libc.src.compiler.__stack_chk_guard
```

Comment:
should this also be enabled for x86_64?

Reply:
x86-64 uses TLS to store the stack canary so it's not needed there.

Reply:
Doesn't that mean then that we //can't// use libc/src/compiler/linux/__stack_chk_guard.cpp for x86?

I guess I have a question around do we need to match existing ABI for different architectures as glibc, or can we do our own ABI? If we do our own ABI, I'd much rather have per thread stack canaries, than a single global for all threads.

Reply:
I think the baremetal impl in this PR is fine.

For the linux target, we need to do more research because I'm quite sure some architectures use TLS for the canary.

Reply:
We already implement support for setting the TLS canary on x86-64 Linux here https://github.com/llvm/llvm-project/blob/ebd4dc42630e201c15894b48c3ea890eaa8c3b18/libc/startup/linux/x86_64/tls.cpp#L67-L76

The location of the stack canary is defined by psABI and if we want to change I think it'd be better to propose a change to psABI for the particular architecture.

Reply:
libc/src/compiler/__stack_chk_guard.h is not specific to baremetal.  Both that header and libc/src/compiler/linux/__stack_chk_guard.cpp declare `__stack_chk_guard` as non-TLS.  Shouldn't they be declaring `__stack_chk_guard` as `__thread` in order to be thread local?

cc @SchrodingerZhu

Reply:
I don't think `__stack_chk_guard` is thread-wise. In MUSL, it is initialized only once with AT_RANDOM. If I don't remember it wrong, then burn threads are copying that value instead of creating new values. I think a function execution can be preempted and transfer to other threads.

Reply:
see also https://github.com/bminor/musl/blob/f314e133929b6379eccc632bef32eaebb66a7335/src/thread/pthread_create.c#L328

Reply:
I might be misunderstanding, but take this example: https://godbolt.org/z/84s39bcsY

`%fs:40` is the thread local slot for the stack canary? i.e. `__stack_chk_guard` ?

Reply:
Yes, it is. `canary` can be either accessed via `__stack_chk_guard` or TLS canary, depending on platform conventions and compiler options. In either case, it is the same value for different threads. Therefore, `__stack_chk_guard` does not need to be `__thread`. My guess is that TLS is used to make it harder to override the canary.

Reply:
Inside linux kernel, I think there are `per-cpu` or even `per-task` canary. But at least for MUSL, I only found a single-valued canary.

Reply:
FYI, TLS and global variable are two styles of protector guards. This is make explicit in RISC-V cmdopts (https://gcc.gnu.org/onlinedocs/gcc/RISC-V-Options.html):
```
-mstack-protector-guard=guard
-mstack-protector-guard-reg=reg
-mstack-protector-guard-offset=offset

    Generate stack protection code using canary at guard. Supported locations are ‘global’ for a global canary or ‘tls’ for per-thread canary in the TLS block.

    With the latter choice the options -mstack-protector-guard-reg=reg and -mstack-protector-guard-offset=offset furthermore specify which register to use as base register for reading the canary, and from what offset from that base register. There is no default register or offset as this is entirely for use within the Linux kernel. 

```